release-common-lib/src/main/java/org/reactome/util/ensembl/EnsemblServiceResponseProcessor.java

```diff
@@ -153,160 +150,249 @@ public EnsemblServiceResponseProcessor()
-					timeoutRetriesRemaining--;
-					logger.error("Request timed out! {} retries remaining", timeoutRetriesRemaining);
-					if (timeoutRetriesRemaining > 0)
-					{
-						okToQuery = true;
-					}
-					else
-					{
-						logger.error("No more retries remaining.");
-						timeoutRetriesRemaining = 3;
-						okToQuery = false;
-					}
-					break;
-				default:
-					// Log any other kind of response.
-					okToQuery = false;
-					try
-					{
-						content = EntityUtils.toString(response.getEntity(), Charset.forName("UTF-8"));
-					}
-					catch (ParseException e)
-					{
-						e.printStackTrace();
-					}
-					catch (IOException e)
-					{
-						e.printStackTrace();
-					}
-					result.setResult(content);
-					logger.info("Unexpected response {} with message: {}",response.getStatusLine().getStatusCode(), response.getStatusLine().getReasonPhrase());
-					break;
-			}
+			logger.warn(
+				"No X-RateLimit-Remaining was returned. This is odd. Response message: {} ; "+
+					"Headers returned are: {} " + System.lineSeparator() +
+					"Last known value for remaining was {}",
+
+				response.getStatusLine().toString(),
+				getHeaders(response),
+				numRequestsRemaining
+			);
 		}
-		result.setOkToRetry(okToQuery);
-		if (response.containsHeader("X-RateLimit-Remaining"))
+	}
+
+	/**
+	 * Returns the duration to wait before retrying the query.  The wait time is determined from the "Retry-After"
+	 * header in the query response and is multiplied by the number of times the query has been requested to wait.
+	 * This is to give the increasing server buffer time on each failure requesting the query waits.
+	 * @param response Response object from which to obtain the base wait time
+	 * @return Wait time as Duration object ("Retry-After" value multiplied by the number of times the query has
+	 * requested to wait).
+	 */
+	private Duration processWaitTime(HttpResponse response)
+	{
+		Duration waitTime = Duration.ofSeconds(parseIntegerHeaderValue(response,"Retry-After"));
+
+		logger.warn("The server told us to wait, so we will wait for {} * {} before trying again.",
+			waitTime, this.waitMultiplier
+		);
+
+		return waitTime.multipliedBy(this.waitMultiplier);
+	}
+
+	/**
+	 * Returns <code>true</code> if the number of times the query the object for this class has been told to wait is
+	 * less than the threshold (set to a maximum of 5); <code>false</code> is returned otherwise and the maximum
+	 * number of times waiting being reached is logged.
+	 *
+	 * The counter for the number of times waited will be incremented each time this method is called when the
+	 * threshold has not been met.
+	 *
+	 * @return <code>true</code> if the number of times query has been told to wait is less than the threshold;
+	 * <code>false</code> otherwise.
+	 */
+	private boolean timesWaitedThresholdNotMet()
+	{
+		// If we get told to wait >= 5 times, let's just take the hint and stop trying.
+		final int MAX_TIMES_TO_WAIT = 5;
```

Comment:
Similar as early on in my PR -- this one is capitalized but should it be globally available?

Reply:
Done!